Mathlib/Algebra/GroupWithZero/Divisibility.lean

```diff
@@ -141,3 +143,66 @@ theorem eq_of_forall_dvd' (h : ∀ c, c ∣ a ↔ c ∣ b) : a = b :=
 #align eq_of_forall_dvd' eq_of_forall_dvd'
 
 end CancelCommMonoidWithZero
+
+namespace Ring
+
+open Classical
+
+variable {M₀ : Type u} [MonoidWithZero M₀]
+
+/-- Introduce a binary function `divide` on monoids with zero `M₀`, which sends `x` and `y` to
+`x / y` if `y` is non-zero and divides `x`, and to `0` otherwise. This definition is somewhat
+ad hoc, but one needs a fully (rather than partially) defined division function for some purposes.
+
+Note that while this is in the `Ring` namespace for brevity, it requires the weaker assumption
+`MonoidWithZero M₀` instead of `Ring M₀`. -/
+noncomputable def divide (x y : M₀) : M₀ :=
+  if h : y ≠ 0 ∧ y ∣ x then h.right.choose else 0
```

Comment:
Why do you require `y` to be non-zero? In a general ring, being non-zero isn't a very natural condition. You can still have plenty of nilpotents and/or zero-divisors.

Reply:
Fair enough. In my head I'm working with only integral domains, but I just didn't want to impose this on the definition. Without this condition I wouldn't be able to prove `divide_zero`, which I think is something nice to have.

Reply:
I mean somthing like epsX / epsX should still be well-defined, I guess?

Reply:
but it'd also be nice to have `zero_divide` in general. I wonder if there's another design that allows this, maybe prioritising zero if it's possible